tests/scripts/analyze_outcomes.py

```diff
@@ -136,11 +136,19 @@ def __init__(self, options) -> None:
     def section_name(self) -> str:
         """The section name to use in results."""
 
-    def is_test_case_ignored(self, full_test_suite: str, test_string: str) -> bool:
+    def ignored_tests(self, test_suite: str) -> typing.Iterator[IgnoreEntry]:
+        """Generate the ignore list for the specified test suite."""
+        if test_suite in self.IGNORED_TESTS:
+            yield from self.IGNORED_TESTS[test_suite]
+        pos = test_suite.find('.')
+        if pos != -1:
```

Comment:
I'm sorry, I don't understand your comment. The code seems correct to me, and it seems to be working in practice, both with test suites containing a dot and test suites not containing a dot.

I don't know what you mean by a “generator bit”. This function doesn't return an iterator, it _is_ an iterator. (A function that returns an iterator would be a function that returns a function that itself has `yield` statements.)

The type annotation does not change the code's behavior. `Iterator[foo]` isn't exactly the type of iterators that yield objects of type foo: it's the return type for functions that are iterators that yield objects of type foo.

I hope this helps, but suspect I'm missing something in your comment.

Reply:
No, sorry, I'm not getting the terminology right. The function _does_ return an [iterator](https://docs.python.org/3/glossary.html#term-iterator) — an object that implements the iterator protocol, i.e. it has a `__next__` method with certain properties. The function itself is a [generator function](https://docs.python.org/3/glossary.html#term-generator).

Reply:
A function that ends in some kind of `yield` is a generator, and will return an iteratable object

Here we have paths that do **not** end in `yield`, instead we just run out of scope. Normally in this case the function would not act as a generator, and would return `None` - this would obviously cause issues for callers expecting this to behave like a generator.

Given the script seems to be functioning correctly, I am presuming that we are saved here by the return type being set, so its converting `None` into an empty iterator, however I am unsure if this is the case, hence wondering if we need to explicitly do that with `yield from ()`, I have never seen a generator function before that did not have a `yield` on all paths, basically.

The 'generator bit' is something set internally to python to indicate that a function return is a generator output, and thus an iterable object, normally if there is no `yield` statement it won't get set, thus the function is not a generator.

Reply:
Ah, ok, I see.

So that's not exactly how generators work.

`yield` does not “end” a generator. The code of the function continues to run after `yield`. It's just paused while the caller consumes the yielded value. If you don't care about the order of executions, `yield foo` is equivalent to `accumulator.append(foo)`, if you additionally do `return accumulator` after the function returns.

Eventually, unless the function loops indefinitely or throws an exception, the function will return. Normally a generator function returns by falling off the end of the function, but you can call `return` explicitly too.

For example, `twice` is a generator that yields its argument twice. `twice_maybe` is a generator that either is empty or yields its argument, flipping a coin each time you call it.

```
def twice(x):
    yield x
    yield x

def twice_maybe(x):
    if random.randint(0, 1):
        return
    yield x
    yield x
```

Under the hood, when the generator function returns, it throws a [`StopIterator`](https://docs.python.org/3/library/exceptions.html#StopIteration) exception. That's how the caller knows that the iterator is finished. Most code that constructs or consumes iterators doesn't mention `StopIteration` explicitly, but it comes up in more complex cases.

A function is a generator function if it contains a [`yield` statement](https://docs.python.org/3/reference/simple_stmts.html#the-yield-statement) (or `yield from`). This is a syntactic property.

In older versions of Python, a function could not both have a `yield` and a `return` with an expression. Not even `return None`, because it's a syntactic constraint.

```
$ cat a.py
def f():
    yield 'a'
    return None
$ python2 a.py
  File "a.py", line 3
    return None
SyntaxError: 'return' with argument inside generator
```

In modern Python, it's possible to have a `return` statement with an expression, but very unusual. You have to handle the `StopIteration` exception to obtain the return value.

You can find some examples of generator functions of varying complexity in the [itertools module documentation](https://docs.python.org/3/library/itertools.html).

Reply:
I understood most of that, but I thought that the function only became a generator once a `yield`  was hit, and thus the above would potentially throw a `StopInteration` exception from the paths not containing a `yield`. It would appear this fear was unfounded.